### PR TITLE
Implement OrderedIterable.getFirstOptional() and OrderedIterable.getL…

### LINKED ISSUE #123: Implement OrderedIterable.getFirstOptional() and OrderedIterable.getLastOptional() .
Implement getFirst and getLast methods that returns Optional on OrderedIterable. The method signatures would look like below.

`Optional<T> getFirstOptional()`
`Optional<T> getLastOptional()`

### PR DESCRIPTION
**Pull request by PascalSchumacher**
…astOptional() (closes #123)

### CODE REVIEW

**Comment by nikhilnanivadekar** [comment]
In order to close #123, we should implement OrderedIterable.maxOptional() and OrderedIterable.minOptional(). It can be a separate commit, but the implementation will touch the same classes.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/ordered/OrderedIterable.java [hunk 1]
```diff
@@ -67,6 +68,17 @@
...
 
     /**
+     * Returns the first element of an iterable as an Optional. In the case of a List it is the element at the first index. In the
+     * case of any other Collection, it is the first element that would be returned during an iteration. If the
+     * iterable is empty, {@link Optional#empty} is returned. If null is a valid element of the container, then a developer would need to
+     * check to see if the iterable is empty to validate that an {@link Optional#empty} result was not due to the container being empty.
+     */
+    default Optional<T> getFirstOptional()
+    {
+        return Optional.ofNullable(getFirst());
```
Comment (nikhilnanivadekar) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/ordered/OrderedIterable.java]: Please qualify with `this`:
`this.getFirst()`

**Comment by motlin** [comment]
@nikhilnanivadekar I'm not sure I understand the point about maxOptional() / minOptional(). That method doesn't exist today, does it?

**Comment by nikhilnanivadekar** [comment]
@motlin you are right. I was thinking about having maxOptional() and minOptional() to be just null safe. If the user has a Comparator with nulls high or nulls low respectively. But I agree with what you are hinting at, and we do not need them.

**Comment by itohro** [comment]
@nikhilnanivadekar @motlin FYI we hane another issue for maxOptional/minOptional in #122 .

**Comment by motlin** [comment]
Right, thanks @itohro. I wasn't hinting at anything @nikhilnanivadekar I just thought it would be a separate piece of work in a separate pull request.

**Comment by nikhilnanivadekar** [comment]
If I understand correctly, this PR will partially close #213.
May be we should clean up #213 . 
Apologies for the very first comment, I got confused between #123 and #213

**Comment by motlin** [comment]
I think that #213 is redundant because the same feature requests are
covered in other issues.

**Comment by motlin** [comment]
Feedback similar to the discussion on #239, and it reminds me of the work @itohro did on detectOptional(). We realized that detectOptional() is more expressive than detect(), so detect() could be rewritten to delegate to detectOptional(), but not the other way around. The problem is that detect() and getFirst() can return null for two reasons, either the found element is null or the no element was found.

If we attempt to go ahead and delegate from getFirstOptional to getFirst in a default implementation, then we at least need a size check, just like the example I wrote on #239. That way, an empty collection would return Optional.empty() and a collection with the first/last element null would throw NullPointerException by attempting Optional.of(null).

**Comment by PascalSchumacher** [comment]
Thanks. Than makes sense. I will update the pull request.

I really should have checked out the implementation of `detectOptional` before submitting this pull request. Sorry for wasting your time.

**Comment by motlin** [comment]
Absolutely not a waste of time. The new patch looks better. It has some CheckStyle violations which I fixed in fe4748a4e65f378c3466e3560b77c4878bab56f3. It's currently [building](https://travis-ci.org/motlin/eclipse-collections/builds/217423357). If the build passes, you could include these changes in your pull request.

**Comment by PascalSchumacher** [comment]
Thanks! I have updated the pull request to include your fixes.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/ordered/OrderedIterable.java [hunk 2]
```diff
@@ -67,6 +68,23 @@
     T getFirst();
 
     /**
+     * Returns the first element of an iterable as an Optional. In the case of a List it is the element at the first index.
+     * In the case of any other Collection, it is the first element that would be returned during an iteration. If the
+     * iterable is empty, {@link Optional#empty} is returned.
+     *
+     * @throws NullPointerException if the element is null<
```
Comment (motlin) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/ordered/OrderedIterable.java]: There's an extra trailing angle bracket on this line.

**Comment by motlin** [comment]
Thank you very much!

**Comment by PascalSchumacher** [comment]
Thanks for merging! :)

Is the since tag I used (`9.0`) correct?

**Comment by nikhilnanivadekar** [comment]
@PascalSchumacher Thanks for bringing it to attention. It should be 8.2.0 which is coming up by the end of April. 
Can you please raise a PR? I will merge it.

**Comment by PascalSchumacher** [comment]
On further consideration: The default methods are broken for synchronized classes. :( There is no guarantee that the collection is still non-empty when `getFirst` is called by `getFirstOptional`.

**Comment by motlin** [comment]
Great point. We have some tests in the scala module that can catch this
sort of problem if we add new tests along with the older ones for getFirst
and getLast.

**Comment by PascalSchumacher** [comment]
I added tests for `getFirstOptional` and `getLastOptional` to `SynchronizedStackScalaTest`, but they are unable to demonstrate a problem. The test have two threads access `getFirstOptional`/`getLastOptional` in parallel and this case safe.

**Pull request merged**